creed/rosetta.py

```diff
@@ -0,0 +1,290 @@
+
+    if batch_size < 1 or not isinstance(batch_size, int):
+        raise ValueError('batch_size parameter must be a positive integer')
+
+    if gaus_rad < 0 or not isinstance(gaus_rad, int):
+        raise ValueError('gaus_rad parameter must be a non-negative integer')
+
+    # loop over each set of FOVs in the batch
+    for i in range(0, len(fovs), batch_size):
+        print("Processing image {}".format(i))
+
+        # load batch of fovs
+        batch_fovs = fovs[i: i + batch_size]
+        batch_data = load_imgs_from_tree(data_dir=raw_data_dir, fovs=batch_fovs,
+                                         channels=acquired_targets, dtype='float32')
+
+        # blur data
+        if gaus_rad > 0:
+            batch_data_blurred = np.zeros_like(batch_data)
+            for j in range(batch_data.shape[0]):
+                for k in range(batch_data.shape[-1]):
+                    blurred = gaussian_filter(batch_data[j, :, :, k], sigma=gaus_rad)
+                    batch_data_blurred[j, :, :, k] = blurred
+        else:
+            batch_data_blurred = batch_data
+
+        comp_data = compensate_matrix_simple(raw_inputs=batch_data_blurred,
+                                             comp_coeffs=comp_mat.values)
+
+        # set negative values to zero
+        comp_data = np.where(comp_data > 0, comp_data, 0)
+
+        # save data
+        for j in range(batch_data.shape[0]):
+            fov_folder = os.path.join(comp_data_dir, batch_data.fovs.values[j])
+            os.makedirs(fov_folder)
+
+            # create directories for saving tifs
+            if save_format in ['normalized', 'both']:
+                norm_folder = os.path.join(fov_folder, 'normalized')
+                os.makedirs(norm_folder)
+
+            if save_format in ['raw', 'both']:
+                raw_folder = os.path.join(fov_folder, 'raw')
+                os.makedirs(raw_folder)
+
+            for k in range(batch_data.shape[-1]):
+                channel_name = batch_data.channels.values[k] + '.tiff'
+
+                # save tifs to appropriate directories
+                if save_format in ['normalized', 'both']:
+                    save_path = os.path.join(norm_folder, channel_name)
+                    io.imsave(save_path, comp_data[j, :, :, k] / norm_const, check_contrast=False)
+
+                if save_format in ['raw', 'both']:
+                    save_path = os.path.join(raw_folder, channel_name)
+                    io.imsave(save_path, comp_data[j, :, :, k], check_contrast=False)
+
+# TODO: Decide if this code is necessary or not
+# def compare_comped_images(raw_dir, comp_dir_list, output_dir):
+#     """Creates a tiled image containing the raw image, difference image, and output image
+#     from each compensated directory supplied
+#
+#     Args:
+#         raw_dir: directory containing raw images
+#         comp_dir_list: list of directories containing compensated images
+#         output_dir: directory where tifs will be saved"""
```

Comment:
similar to the tiling function below, some of this might be useful for other comparisons.  I'd say this can be a function, but maybe the bulk of it's image manipulation can be moved elsewhere (different folder, or part of the future "various utils from ark" package)

Reply:
Yeah, this is basically a carbon copy of the above function, with the added step of creating a difference image between first and second supplied directory. I'm going to remove it, since I think the above function will get the job done, but if Cameron and I end up relying on this heavily then I'll add it back